File: Python/OneLang/Parsers/Common/Reader.py
from OneLangStdLib import *
import OneLang.Parsers.Common.Utils as utils
import re

class Cursor:
    def __init__(self, offset, line, column, line_start, line_end):
        self.offset = offset
        self.line = line
        self.column = column
        self.line_start = line_start
        self.line_end = line_end

class ParseError:
    def __init__(self, message, cursor = None, reader = None):
        self.message = message
        self.cursor = cursor
        self.reader = reader

class Reader:
    def __init__(self, input):
        self.ws_offset = 0
        self.offset = 0
        self.cursor_search = None
        self.line_comment = "//"
        self.supports_block_comment = True
        self.block_comment_start = "/*"
        self.block_comment_end = "*/"
        self.comment_disabled = False
        self.identifier_regex = "[A-Za-z_][A-Za-z0-9_]*"
        self.number_regex = "[+-]?(\\d*\\.\\d+|\\d+\\.\\d+|0x[0-9a-fA-F_]+|0b[01_]+|[0-9_]+)"
        self.errors = []
        self.hooks = None
        self.ws_line_counter = 0
        self.move_ws_offset = True
        self.prev_token_offset = -1
        self.input = input
        self.cursor_search = CursorPositionSearch(input)
    
    def get_eof(self):
        return self.offset >= len(self.input)
    
    def get_cursor(self):
        return self.cursor_search.get_cursor_for_offset(self.offset)
    
    def get_preview(self):
        preview = re.sub("\\n", "\\n", self.input[self.offset:self.offset + 30])
        if len(preview) == 30:
            preview += "..."
        return preview
    
    def line_preview(self, cursor):
        line = self.input[cursor.line_start:cursor.line_end - 1]
        return f'''{line}\n{" " * (cursor.column - 1)}^^^'''
    
    def fail(self, message, offset = -1):
        error = ParseError(message, self.cursor_search.get_cursor_for_offset(self.offset if offset == -1 else offset), self)
        self.errors.append(error)
        
        if self.hooks != None:
            self.hooks.error_callback(error)
        else:
            raise Error(f'''{message} at {error.cursor.line}:{error.cursor.column}\n{self.line_preview(error.cursor)}''')
    
    def skip_whitespace(self, include_in_trivia = False):
        
        while self.offset < len(self.input):
            c = self.input[self.offset]
            
            if c == "\n":
                self.ws_line_counter = self.ws_line_counter + 1
            
            if not (c == "\n" or c == "\r" or c == "\t" or c == " "):
                break
            self.offset = self.offset + 1
        
        if not include_in_trivia:
            self.ws_offset = self.offset
    
    def skip_until(self, token):
        index = self.input.find(token, self.offset)
        if index == -1:
            return False
        self.offset = index + len(token)
        if self.move_ws_offset:
            self.ws_offset = self.offset
        return True
    
    def skip_line(self):
        if not self.skip_until("\n"):
            self.offset = len(self.input)
    
    def is_alpha_num(self, c):
        n = ord(c[0])
        return (97 <= n and n <= 122) or (65 <= n and n <= 90) or (48 <= n and n <= 57) or n == 95
    
    def read_exactly(self, what):
        if self.input.startswith(what, self.offset):
            self.ws_offset = self.offset = self.offset + len(what)
            return True
        return False
    
    def read_char(self):
        # TODO: should we move wsOffset?
        self.offset = self.offset + 1
        return self.input[self.offset - 1]
    
    def peek_token(self, token):
        self.skip_whitespace_and_comment()
        
        if self.input.startswith(token, self.offset):
            # TODO: hackish way to make sure space comes after word tokens
            if self.is_alpha_num(token[len(token) - 1]) and self.is_alpha_num(self.input[self.offset + len(token)]):
                return False
            return True
        else:
            return False
    
    def read_token(self, token):
        if self.peek_token(token):
            self.prev_token_offset = self.offset
            self.ws_offset = self.offset = self.offset + len(token)
            return True
        return False
    
    def read_any_of(self, tokens):
        for token in tokens:
            if self.read_token(token):
                return token
        return None
    
    def expect_token(self, token, error_msg = None):
        if not self.read_token(token):
            self.fail(error_msg or f'''expected token \'{token}\'''')
    
    def expect_string(self, error_msg = None):
        result = self.read_string()
        if result == None:
            self.fail(error_msg or f'''expected string''')
        return result
    
    def expect_one_of(self, tokens):
        result = self.read_any_of(tokens)
        if result == None:
            self.fail(f'''expected one of the following tokens: {", ".join(tokens)}''')
        return result
    
    @classmethod
    def match_from_index(cls, pattern, input, offset):
        regex = RegExp(pattern, "gy")
        regex.last_index = offset
        matches = regex.exec(input)
        if matches == None:
            return None
        else:
            result = []
            i = 0
            
            while i < len(matches):
                result.append(matches[i])
                i = i + 1
            return result
    
    def peek_regex(self, pattern):
        matches = Reader.match_from_index(pattern, self.input, self.offset)
        return matches
    
    def read_regex(self, pattern):
        matches = Reader.match_from_index(pattern, self.input, self.offset)
        if matches != None:
            self.prev_token_offset = self.offset
            self.ws_offset = self.offset = self.offset + len(matches[0])
        return matches
    
    def skip_whitespace_and_comment(self):
        if self.comment_disabled:
            return
        
        self.move_ws_offset = False
        while True:
            self.skip_whitespace(True)
            if self.input.startswith(self.line_comment, self.offset):
                self.skip_line()
            elif self.supports_block_comment and self.input.startswith(self.block_comment_start, self.offset):
                if not self.skip_until(self.block_comment_end):
                    self.fail(f'''block comment end ("{self.block_comment_end}") was not found''')
            else:
                break
        self.move_ws_offset = True
    
    def read_leading_trivia(self):
        self.skip_whitespace_and_comment()
        this_line_start = self.input.rfind("\n", 0, self.offset)
        if this_line_start <= self.ws_offset:
            return ""
        
        result = self.input[self.ws_offset:this_line_start + 1]
        result = utils.Utils.deindent(result)
        self.ws_offset = this_line_start
        return result
    
    def read_identifier(self):
        self.skip_whitespace()
        id_match = self.read_regex(self.identifier_regex)
        if id_match == None:
            return None
        
        return id_match[0]
    
    def read_number(self):
        self.skip_whitespace()
        num_match = self.read_regex(self.number_regex)
        if num_match == None:
            return None
        
        if self.read_regex("[0-9a-zA-Z]") != None:
            self.fail("invalid character in number")
        
        return num_match[0]
    
    def read_string(self):
        self.skip_whitespace()
        
        sep_char = self.input[self.offset]
        if sep_char != "'" and sep_char != "\"":
            return None
        
        str = ""
        self.read_exactly(sep_char)
        while not self.read_exactly(sep_char):
            chr = self.read_char()
            if chr == "\\":
                esc = self.read_char()
                if esc == "n":
                    str += "\n"
                elif esc == "r":
                    str += "\r"
                elif esc == "t":
                    str += "\t"
                elif esc == "\\":
                    str += "\\"
                elif esc == sep_char:
                    str += sep_char
                else:
                    self.fail("invalid escape", self.offset - 1)
            else:
                chr_code = ord(chr[0])
                if not (32 <= chr_code and chr_code <= 126) or chr == "\\" or chr == sep_char:
                    self.fail(f'''not allowed character (code={chr_code})''', self.offset - 1)
                str += chr
        return str
    
    def expect_identifier(self, error_msg = None):
        id = self.read_identifier()
        if id == None:
            self.fail(error_msg or "expected identifier")
        return id
    
    def read_modifiers(self, modifiers):
        result = []
        while True:
            success = False
            for modifier in modifiers:
                if self.read_token(modifier):
                    result.append(modifier)
                    success = True
            if not success:
                break
        return result

class CursorPositionSearch:
    def __init__(self, input):
        self.line_offsets = [0]
        self.input = input
        i = 0
        
        while i < len(input):
            if input[i] == "\n":
                self.line_offsets.append(i + 1)
            i = i + 1
        self.line_offsets.append(len(input))
    
    def get_line_idx_for_offset(self, offset):
        low = 0
        high = len(self.line_offsets) - 1
        
        while low <= high:
            # @java var middle = (int)Math.floor((low + high) / 2);
            middle = Math.floor((low + high) / 2)
            middle_offset = self.line_offsets[middle]
            if offset == middle_offset:
                return middle
            elif offset <= middle_offset:
                high = middle - 1
            else:
                low = middle + 1
        
        return low - 1
    
    def get_cursor_for_offset(self, offset):
        line_idx = self.get_line_idx_for_offset(offset)
        line_start = self.line_offsets[line_idx]
        line_end = self.line_offsets[line_idx + 1]
        column = offset - line_start + 1
        if column < 1:
            raise Error("Column should not be < 1")
        return Cursor(offset, line_idx + 1, offset - line_start + 1, line_start, line_end)

File: Java/src/main/java/OneLang/Parsers/Common/Reader/Reader.java
package OneLang.Parsers.Common.Reader;

import OneLang.Parsers.Common.Utils.Utils;

import OneLang.Parsers.Common.Reader.CursorPositionSearch;
import java.util.List;
import OneLang.Parsers.Common.Reader.ParseError;
import OneLang.Parsers.Common.Reader.IReaderHooks;
import OneLang.Parsers.Common.Reader.Cursor;
import OneStd.RegExp;
import java.util.regex.Pattern;
import java.util.ArrayList;
import OneStd.Objects;
import java.util.stream.Collectors;
import java.util.Arrays;

public class Reader {
    public Integer wsOffset = 0;
    public Integer offset = 0;
    public CursorPositionSearch cursorSearch;
    public String lineComment = "//";
    public Boolean supportsBlockComment = true;
    public String blockCommentStart = "/*";
    public String blockCommentEnd = "*/";
    public Boolean commentDisabled = false;
    public String identifierRegex = "[A-Za-z_][A-Za-z0-9_]*";
    public String numberRegex = "[+-]?(\\d*\\.\\d+|\\d+\\.\\d+|0x[0-9a-fA-F_]+|0b[01_]+|[0-9_]+)";
    public List<ParseError> errors;
    public IReaderHooks hooks;
    public Integer wsLineCounter = 0;
    public Boolean moveWsOffset = true;
    public Integer prevTokenOffset;
    public String input;
    
    public Boolean getEof() {
        return this.offset >= this.input.length();
    }
    
    public Cursor getCursor() {
        return this.cursorSearch.getCursorForOffset(this.offset);
    }
    
    public String getPreview() {
        var preview = this.input.substring(this.offset, this.offset + 30).replaceAll("\\n", "\\n");
        if (preview.length() == 30)
            preview += "...";
        return preview;
    }
    
    public Reader(String input)
    {
        this.input = input;
        this.errors = new ArrayList<ParseError>();
        this.hooks = null;
        this.prevTokenOffset = -1;
        this.cursorSearch = new CursorPositionSearch(input);
    }
    
    public String linePreview(Cursor cursor) {
        var line = this.input.substring(cursor.lineStart, cursor.lineEnd - 1);
        return line + "\n" + " ".repeat(cursor.column - 1) + "^^^";
    }
    
    public void fail(String message, Integer offset) {
        var error = new ParseError(message, this.cursorSearch.getCursorForOffset(offset == -1 ? this.offset : offset), this);
        this.errors.add(error);
        
        if (this.hooks != null)
            this.hooks.errorCallback(error);
        else
            throw new Error(message + " at " + error.cursor.line + ":" + error.cursor.column + "\n" + this.linePreview(error.cursor));
    }
    
    public void skipWhitespace(Boolean includeInTrivia) {
        for (; this.offset < this.input.length(); this.offset++) {
            var c = this.input.substring(this.offset, this.offset + 1);
            
            if (Objects.equals(c, "\n"))
                this.wsLineCounter++;
            
            if (!(Objects.equals(c, "\n") || Objects.equals(c, "\r") || Objects.equals(c, "\t") || Objects.equals(c, " ")))
                break;
        }
        
        if (!includeInTrivia)
            this.wsOffset = this.offset;
    }
    
    public Boolean skipUntil(String token) {
        var index = this.input.indexOf(token, this.offset);
        if (index == -1)
            return false;
        this.offset = index + token.length();
        if (this.moveWsOffset)
            this.wsOffset = this.offset;
        return true;
    }
    
    public void skipLine() {
        if (!this.skipUntil("\n"))
            this.offset = this.input.length();
    }
    
    public Boolean isAlphaNum(String c) {
        var n = (int)c.charAt(0);
        return (97 <= n && n <= 122) || (65 <= n && n <= 90) || (48 <= n && n <= 57) || n == 95;
    }
    
    public Boolean readExactly(String what) {
        if (this.input.startsWith(what, this.offset)) {
            this.wsOffset = this.offset = this.offset + what.length();
            return true;
        }
        return false;
    }
    
    public String readChar() {
        // TODO: should we move wsOffset?
        this.offset++;
        return this.input.substring(this.offset - 1, this.offset - 1 + 1);
    }
    
    public Boolean peekToken(String token) {
        this.skipWhitespaceAndComment();
        
        if (this.input.startsWith(token, this.offset)) {
            // TODO: hackish way to make sure space comes after word tokens
            if (this.isAlphaNum(token.substring(token.length() - 1, token.length() - 1 + 1)) && this.isAlphaNum(this.input.substring(this.offset + token.length(), this.offset + token.length() + 1)))
                return false;
            return true;
        }
        else
            return false;
    }
    
    public Boolean readToken(String token) {
        if (this.peekToken(token)) {
            this.prevTokenOffset = this.offset;
            this.wsOffset = this.offset = this.offset + token.length();
            return true;
        }
        return false;
    }
    
    public String readAnyOf(String[] tokens) {
        for (var token : tokens) {
            if (this.readToken(token))
                return token;
        }
        return null;
    }
    
    public void expectToken(String token, String errorMsg) {
        if (!this.readToken(token)) {
            var result = errorMsg;
            this.fail(result != null ? result : "expected token '" + token + "'", -1);
        }
    }
    
    public String expectString(String errorMsg) {
        var result = this.readString();
        if (result == null) {
            var result2 = errorMsg;
            this.fail(result2 != null ? result2 : "expected string", -1);
        }
        return result;
    }
    
    public String expectOneOf(String[] tokens) {
        var result = this.readAnyOf(tokens);
        if (result == null)
            this.fail("expected one of the following tokens: " + Arrays.stream(tokens).collect(Collectors.joining(", ")), -1);
        return result;
    }
    
    public static String[] matchFromIndex(String pattern, String input, Integer offset) {
        var regex = new RegExp(pattern, "gy");
        regex.lastIndex = offset;
        var matches = regex.exec(input);
        if (matches == null)
            return null;
        else {
            var result = new ArrayList<String>();
            for (Integer i = 0; i < matches.length; i++)
                result.add(matches[i]);
            return result.toArray(String[]::new);
        }
    }
    
    public String[] peekRegex(String pattern) {
        var matches = Reader.matchFromIndex(pattern, this.input, this.offset);
        return matches;
    }
    
    public String[] readRegex(String pattern) {
        var matches = Reader.matchFromIndex(pattern, this.input, this.offset);
        if (matches != null) {
            this.prevTokenOffset = this.offset;
            this.wsOffset = this.offset = this.offset + matches[0].length();
        }
        return matches;
    }
    
    public void skipWhitespaceAndComment() {
        if (this.commentDisabled)
            return;
        
        this.moveWsOffset = false;
        while (true) {
            this.skipWhitespace(true);
            if (this.input.startsWith(this.lineComment, this.offset))
                this.skipLine();
            else if (this.supportsBlockComment && this.input.startsWith(this.blockCommentStart, this.offset)) {
                if (!this.skipUntil(this.blockCommentEnd))
                    this.fail("block comment end (\"" + this.blockCommentEnd + "\") was not found", -1);
            }
            else
                break;
        }
        this.moveWsOffset = true;
    }
    
    public String readLeadingTrivia() {
        this.skipWhitespaceAndComment();
        var thisLineStart = this.input.lastIndexOf("\n", this.offset);
        if (thisLineStart <= this.wsOffset)
            return "";
        
        var result = this.input.substring(this.wsOffset, thisLineStart + 1);
        result = Utils.deindent(result);
        this.wsOffset = thisLineStart;
        return result;
    }
    
    public String readIdentifier() {
        this.skipWhitespace(false);
        var idMatch = this.readRegex(this.identifierRegex);
        if (idMatch == null)
            return null;
        
        return idMatch[0];
    }
    
    public String readNumber() {
        this.skipWhitespace(false);
        var numMatch = this.readRegex(this.numberRegex);
        if (numMatch == null)
            return null;
        
        if (this.readRegex("[0-9a-zA-Z]") != null)
            this.fail("invalid character in number", -1);
        
        return numMatch[0];
    }
    
    public String readString() {
        this.skipWhitespace(false);
        
        var sepChar = this.input.substring(this.offset, this.offset + 1);
        if (!Objects.equals(sepChar, "'") && !Objects.equals(sepChar, "\""))
            return null;
        
        var str = "";
        this.readExactly(sepChar);
        while (!this.readExactly(sepChar)) {
            var chr = this.readChar();
            if (Objects.equals(chr, "\\")) {
                var esc = this.readChar();
                if (Objects.equals(esc, "n"))
                    str += "\n";
                else if (Objects.equals(esc, "r"))
                    str += "\r";
                else if (Objects.equals(esc, "t"))
                    str += "\t";
                else if (Objects.equals(esc, "\\"))
                    str += "\\";
                else if (Objects.equals(esc, sepChar))
                    str += sepChar;
                else
                    this.fail("invalid escape", this.offset - 1);
            }
            else {
                var chrCode = (int)chr.charAt(0);
                if (!(32 <= chrCode && chrCode <= 126) || Objects.equals(chr, "\\") || Objects.equals(chr, sepChar))
                    this.fail("not allowed character (code=" + chrCode + ")", this.offset - 1);
                str += chr;
            }
        }
        return str;
    }
    
    public String expectIdentifier(String errorMsg) {
        var id = this.readIdentifier();
        if (id == null) {
            var result = errorMsg;
            this.fail(result != null ? result : "expected identifier", -1);
        }
        return id;
    }
    
    public String[] readModifiers(String[] modifiers) {
        var result = new ArrayList<String>();
        while (true) {
            var success = false;
            for (var modifier : modifiers) {
                if (this.readToken(modifier)) {
                    result.add(modifier);
                    success = true;
                }
            }
            if (!success)
                break;
        }
        return result.toArray(String[]::new);
    }
}